endpoints/openrtb2/video_auction.go

```diff
@@ -0,0 +1,523 @@
+		return
+	}
+
+	ctx := context.Background()
+	cancel := func() {}
+	timeout := deps.cfg.AuctionTimeouts.LimitAuctionTimeout(time.Duration(bidReq.TMax) * time.Millisecond)
+	if timeout > 0 {
+		ctx, cancel = context.WithDeadline(ctx, start.Add(timeout))
+	}
+	defer cancel()
+
+	usersyncs := usersync.ParsePBSCookieFromRequest(r, &(deps.cfg.HostCookie))
+	if bidReq.App != nil {
+		labels.Source = pbsmetrics.DemandApp
+	} else {
+		labels.Source = pbsmetrics.DemandWeb
+		if usersyncs.LiveSyncCount() == 0 {
+			labels.CookieFlag = pbsmetrics.CookieFlagNo
+		} else {
+			labels.CookieFlag = pbsmetrics.CookieFlagYes
+		}
+	}
+
+	numImps = len(bidReq.Imp)
+
+	//execute auction logic
+	response, err := deps.ex.HoldAuction(ctx, bidReq, usersyncs, labels, &deps.categories)
+	ao.Request = bidReq
+	ao.Response = response
+	if err != nil {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+
+	//build simplified response
+	bidResp, err := buildVideoResponse(response)
+	if err != nil {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+	if bidReq.Test == 1 {
+		bidResp.Ext = response.Ext
+	}
+
+	resp, err := json.Marshal(bidResp)
+	//resp, err := json.Marshal(response)
+	if err != nil {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(resp)
+
+}
+
+func handleError(labels pbsmetrics.Labels, w http.ResponseWriter, errL []error, ao analytics.AuctionObject) {
+	labels.RequestStatus = pbsmetrics.RequestStatusErr
+	w.WriteHeader(http.StatusInternalServerError)
+	var errors string
+	for _, er := range errL {
+		errors = fmt.Sprintf("%s %s", errors, er.Error())
+	}
+	fmt.Fprintf(w, "Critical error while running the video endpoint: %v", errors)
+	glog.Errorf("/openrtb2/video Critical error: %v", errors)
+	ao.Status = http.StatusInternalServerError
+	ao.Errors = append(ao.Errors, errL...)
+}
+
+func (deps *endpointDeps) createImpressions(videoReq *openrtb_ext.BidRequestVideo) (imps []openrtb.Imp, err []error) {
+	videoDur := videoReq.PodConfig.DurationRangeSec
+	minDuration, maxDuration := minMax(videoDur)
+	reqExactDur := videoReq.PodConfig.RequireExactDuration
+	videoData := videoReq.Video
+
+	finalImpsArray := make([]openrtb.Imp, 0)
+	for _, pod := range videoReq.PodConfig.Pods {
+
+		//load stored impression
+		storedImpressionId := string(pod.ConfigId)
+		storedImp, errs := deps.loadStoredImp(storedImpressionId)
+		if errs != nil {
+			return nil, errs
+		}
+
+		numImps := pod.AdPodDurationSec / minDuration
+
+		impDivNumber := numImps / len(videoDur)
+
+		impsArray := make([]openrtb.Imp, numImps)
+		for impInd := range impsArray {
+			newImp := createImpressionTemplate(storedImp, videoData)
+			impsArray[impInd] = newImp
+			if reqExactDur {
+				//floor := int(math.Floor(ind/impDivNumber))
+				durationIndex := impInd / impDivNumber
+				if durationIndex > len(videoDur)-1 {
+					durationIndex = len(videoDur) - 1
+				}
```

Comment:
Above comment about dividing by zero arose while trying to determine whether above three lines are ever necessary.  As is, I don't think they are, but adding the special case suggested above might change that, though I suspect it doesn't.

Reply:
@josephveach this is a handler for special case we we have let say 7 imps and 2 durations:[15,30]. we cannot divide 7 by 2, so with this code we will have first 3 imps duration equal to 15 and other 4 imps duration equals to 30.

Reply:
My point here was that I don't think it's mathematically possible for durationIndex to ever exceed (len(videoDur)-1.  It's not a big deal.  I was just trying to eliminate an unnecessary check.

Reply:
@josephveach this is possible for example if there are 7 imps and durations array [15, 30] - size 
 2. 
7/2=3, 3 - is incorrect index, we don't have it in array of durations.